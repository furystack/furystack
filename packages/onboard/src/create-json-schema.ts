import { writeFile } from 'fs'
import { createGenerator } from 'ts-json-schema-generator'

export const createJsonSchema = async (options: {
  schemaName: string
  inputTsFile: string
  rootType: string
  outputJsonFile: string
}) => {
  const p = createGenerator({
    path: options.inputTsFile,
    expose: 'export',
    jsDoc: 'extended',
    skipTypeCheck: false,
    topRef: true,
    type: options.rootType,
  })
  const schema = p.createSchema(options.schemaName)
  return await new Promise<typeof schema>((resolve, reject) =>
    writeFile(options.outputJsonFile, JSON.stringify(schema), err => {
      err ? reject(err) : resolve(schema)
    }),
  )
}
