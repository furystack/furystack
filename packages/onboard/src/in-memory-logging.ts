import { writeFileSync } from 'fs'
import { join } from 'path'
import { AbstractLogger, LeveledLogEntry } from '@furystack/logging'
import { Injectable } from '@furystack/inject'

@Injectable({ lifetime: 'singleton' })
export class InMemoryLogging extends AbstractLogger {
  public readonly entries: Array<LeveledLogEntry<any> & { added: Date }> = []

  public async addEntry<T>(entry: import('@furystack/logging').LeveledLogEntry<T>) {
    this.entries.push({ ...entry, added: new Date() })
  }

  public flushToFile(path: string = process.cwd(), logName = 'onboard.log') {
    writeFileSync(join(path, logName), JSON.stringify(this.entries, undefined, 2))
  }
}
