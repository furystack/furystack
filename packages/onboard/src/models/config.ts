import { ServiceModel } from './service'

/**
 * Configuration Root Type
 */
export interface Config {
  $schema: string
  /**
   * Absolute paths of the specific directories
   */
  directories: {
    /**
     * Will be used as the root of your cloned services
     */
    output: string

    /**
     * You can place input data, DB dumps, etc... in this folder
     */
    input: string
  }
  /**
   * List of service definitions
   */
  services: ServiceModel[]
}
