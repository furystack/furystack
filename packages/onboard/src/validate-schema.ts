import Ajv from 'ajv'
import { JSONSchema7 } from 'json-schema'

export const validateSchema = <T>(options: { data: T; schema: JSONSchema7 }) => {
  const ajv = new Ajv()
  const validator = ajv.compile(options.schema)
  const result = validator(options.data)

  return { result, errors: validator.errors }
}
