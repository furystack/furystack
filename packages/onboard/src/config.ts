import { join, dirname } from 'path'
import { readFileSync, realpathSync, existsSync } from 'fs'
import { Injector } from '@furystack/inject/dist/injector'
import { validateSchema } from './validate-schema'
import { Config as ConfigModel } from './models/config'
import { Prerequisite } from './services/check-prerequisites'
import { createJsonSchema } from './create-json-schema'
import { InstallStep } from './models/install-step'

export const configSchemaPath = join(dirname(realpathSync(__filename)), '../config-schema.json')

export interface ConfigOptions {
  workingDir: string
  configSource: string
  parallel: number
  stepFilters?: Array<InstallStep['type']>
  services?: string[]
}

export class Config {
  private configData?: ConfigModel
  public getConfigData = () => {
    if (!this.configData) {
      throw new Error('Config not initialized yet!')
    }
    return { ...this.configData }
  }

  public prerequisites: Prerequisite[] = [
    async (injector) => {
      const outputDir = injector.getConfig().getConfigData().directories.output
      if (!existsSync(outputDir)) {
        return { message: `The output directory '${outputDir}' doesn't exists.`, success: false }
      }
    },
    async (injector) => {
      const inputDir = injector.getConfig().getConfigData().directories.input
      if (!existsSync(inputDir)) {
        return { message: `The input directory '${inputDir}' doesn't exists.`, success: false }
      }
    },
  ]

  public getConfigFilePath = () => join(this.options.workingDir, this.options.configSource)

  public async init() {
    if (!existsSync(configSchemaPath)) {
      await createJsonSchema({
        inputTsFile: join(realpathSync(dirname(__filename)), '..', './src/models/config.ts'),
        outputJsonFile: configSchemaPath,
        rootType: 'Config',
        schemaName: 'Config',
      })
    }
    const configFilePath = this.getConfigFilePath()
    const configFileString = readFileSync(configFilePath)
    const configData = JSON.parse(configFileString.toString()) as ConfigModel

    const schema = JSON.parse(readFileSync(configSchemaPath).toString())
    const validate = validateSchema({ data: configData, schema })
    if (validate.errors) {
      throw Error(
        `${validate.errors.map((e) => `${e.message} - ${e.dataPath} / ${JSON.stringify(e.params)}`).join('\r\n')}`,
      )
    }
    this.configData = configData
  }
  constructor(public readonly options: ConfigOptions) {}
}

declare module '@furystack/inject/dist/injector' {
  export interface Injector {
    useConfig: (config: ConfigOptions) => Injector
    getConfig(): Config
  }
}

Injector.prototype.useConfig = function (cfg) {
  this.setExplicitInstance(new Config(cfg), Config)
  return this
}

Injector.prototype.getConfig = function () {
  return this.getInstance(Config)
}
