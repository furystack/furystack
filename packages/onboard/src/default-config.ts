import { Config } from './models/config'

export const defaultConfig: Config = {
  services: [],
  directories: { output: '~/CradleOutTemp', input: '~/CradleInput' },
}
