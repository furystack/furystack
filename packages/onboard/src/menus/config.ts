import { existsSync, writeFileSync, unlinkSync } from 'fs'
import { spawnSync } from 'child_process'
import { defaultConfig } from '../default-config'
import { terminal } from 'terminal-kit'
import { Injector } from '@furystack/inject'

const getCommandLine = () => {
  switch (process.platform) {
    case 'darwin':
      return 'open'
    case 'win32':
      return 'start'
    default:
      return 'xdg-open'
  }
}

export const openCfgFile = (cfgFile: string) => {
  try {
    return spawnSync(getCommandLine(), [cfgFile], { stdio: 'ignore' })
  } catch (error) {
    terminal.red(error.toString())
  }
}

export const configMenu = async (injector: Injector) => {
  const cfgFile = injector.getConfig().getConfigFilePath()
  const hasConfig = existsSync(cfgFile)
  terminal.saveCursor()

  try {
    await injector.getConfig().init()
  } catch (error) {
    terminal
      .nextLine(1)
      .red('There was an error reading / parsing the current Config file')
      .nextLine(2)
      .red(error.toString())
      .nextLine(1)
  }

  const result = await terminal.singleColumnMenu([
    hasConfig ? 'Edit Config file' : 'Create and edit custom config',
    ...(hasConfig ? ['Reset to defaults', 'Remove custom config'] : []),
    'Back',
  ]).promise

  switch (result.selectedText) {
    case 'Create and edit custom config':
      writeFileSync(cfgFile, JSON.stringify(defaultConfig, undefined, 2))
      openCfgFile(cfgFile)
      break
    case 'Remove custom config':
      unlinkSync(cfgFile)
      terminal.nextLine(2).defaultColor('Config file removed.')
      break
    case 'Reset to defaults':
      writeFileSync(cfgFile, JSON.stringify(defaultConfig, undefined, 2))
      break
    case 'Edit Config file':
      openCfgFile(cfgFile)
      break
    default:
      break
  }

  terminal.restoreCursor()
  terminal.eraseDisplayBelow()
}
