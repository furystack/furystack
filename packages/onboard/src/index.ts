import { writeFileSync } from 'fs'
import { join } from 'path'
import './config'
import yargs from 'yargs'
import { Injector } from '@furystack/inject'
import { ConsoleLogger, VerboseConsoleLogger } from '@furystack/logging'
import got from 'got'
import { mainMenu } from './menus/main'
import { InMemoryLogging } from './in-memory-logging'
import { CheckPrerequisitesService, genericPrerequisites } from './services/check-prerequisites'
import { defaultConfig } from './default-config'
import { installAllServices } from './install-steps/install-all-services'
import { InstallStep } from './models/install-step'
const injector = new Injector().useLogging(InMemoryLogging)

export interface ArgType {
  'download-config': string
  config: string
  parallel: number
  stepFilters?: Array<InstallStep['type']>
  services?: string[]
  verbose?: boolean
}

const initConfig = async (args: ArgType) => {
  args.verbose ? injector.useLogging(VerboseConsoleLogger) : injector.useLogging(ConsoleLogger)
  const logger = injector.logger.withScope('InitConfig')
  if (args['download-config']) {
    logger.information({
      message: 'Downloading remote config...',
      data: { download: args.config, config: args.config },
    })
    const result = await got(args['download-config'])
    writeFileSync(args.config, result.body)
  }
  injector.useConfig({
    configSource: args.config as string,
    workingDir: process.cwd(),
    stepFilters: args.stepFilters,
    services: args.services,
    parallel: args.parallel,
  })

  await injector.getConfig().init()

  const prereqChecks = await injector
    .getInstance(CheckPrerequisitesService)
    .check(...injector.getConfig().prerequisites, ...genericPrerequisites)

  if (prereqChecks.length) {
    logger.error({ message: `The prerequisites has not been met`, data: prereqChecks })
    throw Error(`The prerequisites has not been met`)
  }
}

const cmd = yargs
  .scriptName('onboard')
  .showHelpOnFail(false, 'Specify --help for available options')
  .command(
    'init',
    'initializes the onboard service with a default config file',
    () => {
      /** */
    },
    (args) => {
      try {
        const cfg = JSON.stringify(defaultConfig, undefined, 2)
        writeFileSync(join(process.cwd(), args.config as string), cfg)
      } catch (error) {
        injector.getInstance(InMemoryLogging).flushToFile()
      }
    },
  )
  .command(
    'start',
    'start the selected installations (all services without the --service option)',
    () => {
      /** */
    },
    async (args) => {
      try {
        await initConfig(args as any)
        await installAllServices(injector, injector.getConfig().options.stepFilters)
      } catch (error) {
        injector.getInstance(InMemoryLogging).flushToFile()
      }
      process.exit(0)
    },
  )
  .command(
    '*',
    'starts the ui',
    () => {
      /** */
    },
    async (args) => {
      let run = true
      while (run) {
        try {
          await initConfig(args as any)
          await mainMenu(injector)
        } catch (error) {
          run = false
          injector.logger.error({
            scope: '@furystack/onboard',
            message: `Operation failed: ${error.toString()}`,
            data: { error },
          })
          injector.getInstance(InMemoryLogging).flushToFile()
          process.exit(1)
        }
      }
    },
  )
  .option('verbose', {
    alias: 'v',
    type: 'boolean',
    description: 'Run with verbose logging',
  })
  .option('config', {
    alias: 'cfg',
    type: 'string',
    description: 'The config file path (relative to CWD)',
    default: './onboard-config.json',
  })
  .option('download-config', {
    alias: 'dlc',
    type: 'string',
    description: 'downloads a config file from a specified URL',
  })
  .option('parallel', {
    type: 'number',
    description: 'how many installs can run parallelly',
    default: 1,
  })
  .option('services', {
    type: 'array',
    description: 'a list of services to install',
  })
  .option('stepFilters', {
    type: 'array',
    desciption:
      'a list of types of steps to execute (e.g. if you only want to exec GIT pull(s) and PM2 ADD(s) on each repo',
  })

export default cmd.argv
