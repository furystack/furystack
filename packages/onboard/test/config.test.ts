import { existsSync, unlinkSync, realpathSync } from 'fs'
import { join } from 'path'
import { using, usingAsync } from '@furystack/utils'
import { Injector } from '@furystack/inject'
import { Config } from '../src/config'
import { defaultConfig } from '../src/default-config'

describe('Onboard Config', () => {
  it('Default config should match the snapshot', () => {
    expect(defaultConfig).toMatchSnapshot()
  })

  it('Should extend the Injector with useConfig()', () => {
    using(new Injector(), i => {
      i.useConfig({
        configSource: '',
        parallel: 1,
        workingDir: process.cwd(),
      })
      expect(i.getConfig()).toBeInstanceOf(Config)
    })
  })

  it('Should throw if try to get config data before initialization', () => {
    usingAsync(new Injector(), async i => {
      const cfg = i
        .useConfig({ parallel: 1, configSource: 'onboard-config.json', workingDir: process.cwd() })
        .getConfig()

      expect(() => cfg.getConfigData()).toThrowError()
    })
  })

  it('Should be initialized with a default value', async () => {
    await usingAsync(new Injector(), async i => {
      await i
        .useConfig({
          parallel: 1,
          configSource: 'onboard-config.json',
          workingDir: join(__dirname, '..'),
        })
        .getConfig()
        .init()
      expect(i.getConfig().getConfigData()).toMatchSnapshot()
    })
  })

  it('Should re-create the JSON Schema file during init if not exists', async () => {
    const schemaPath = join(__dirname, '..', 'config-schema.json')
    if (existsSync(schemaPath)) {
      unlinkSync(schemaPath)
    }
    await usingAsync(new Injector(), async i => {
      await i
        .useConfig({
          parallel: 1,
          configSource: 'onboard-config.json',
          workingDir: join(realpathSync(__dirname), '..'),
        })
        .getConfig()
        .init()
      expect(existsSync(schemaPath))
    })
  })
})
