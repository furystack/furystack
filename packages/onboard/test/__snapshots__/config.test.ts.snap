// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Onboard Config Default config should match the snapshot 1`] = `
Object {
  "$schema": "https://gist.github.com/gallayl/38b1a1c279ad8ff192e3902ba020f101/raw",
  "directories": Object {
    "input": "~/dev/onboard-input",
    "output": "~/dev/onboard-output",
  },
  "services": Array [],
}
`;

exports[`Onboard Config Should be initialized with a default value 1`] = `
Object {
  "$schema": "https://gist.github.com/gallayl/38b1a1c279ad8ff192e3902ba020f101/raw",
  "directories": Object {
    "input": "~/dev/onboard-input",
    "output": "~/dev/onboard-output",
  },
  "services": Array [],
}
`;
